src/ocf_composite_volume.c

```diff
@@ -0,0 +1,357 @@
+
+	if (bytes == 0) {
+		/* Flush io - allocate io for each volume */
+		for (i = 0; i < composite->members_cnt; i++) {
+			cio->io[i] = ocf_io_new(&composite->member[i].volume,
+					queue, 0, 0, dir, 0, 0);
+			if (!cio->io[i])
+				goto err;
+		}
+		cio->begin_member = 0;
+		cio->end_member = composite->members_cnt;
+
+		return ioi;
+	}
+
+	for (i = 0; i < composite->members_cnt; i++) {
+		if (addr < composite->lookup[i]) {
+			cio->begin_member = i;
+			break;
+		}
+	}
+
+	for (; i < composite->members_cnt; i++) {
+		begin = i > 0 ? composite->lookup[i - 1] : 0;
+		end = composite->lookup[i];
+
+		io_addr = addr > begin ? addr - begin : 0;
+		io_bytes = OCF_MIN(bytes, end - addr);
+		cio->io[i] = ocf_io_new(&composite->member[i].volume, queue,
+				io_addr, io_bytes, dir, 0, 0);
+		if (!cio->io[i])
+			goto err;
+
+		if (addr + bytes <= end) {
+			cio->end_member = i + 1;
+			break;
+		}
+	}
+
+	return ioi;
+
+err:
+	for (i = 0; i < composite->members_cnt; i++) {
+		if (cio->io[i])
+			ocf_io_put(cio->io[i]);
+	}
+
+	ocf_io_allocator_default_del(allocator, ioi);
+
+	return NULL;
+}
+
+static void ocf_composite_io_allocator_del(ocf_io_allocator_t allocator, void *obj)
+{
+	struct ocf_io_internal *ioi = obj;
+	struct ocf_composite_volume_io *cio = ocf_io_get_priv(&ioi->io);
+	int i;
+
+	for (i = cio->begin_member; i < cio->end_member; i++) {
+		if (cio->io[i])
+			ocf_io_put(cio->io[i]);
+	}
+
+	ocf_io_allocator_default_del(allocator, ioi);
+}
+
+const struct ocf_io_allocator_type ocf_composite_io_allocator_type = {
+	.ops = {
+		.allocator_init = ocf_composite_io_allocator_init,
+		.allocator_deinit = ocf_composite_io_allocator_deinit,
+		.allocator_new = ocf_composite_io_allocator_new,
+		.allocator_del = ocf_composite_io_allocator_del,
+	},
+};
+
+const struct ocf_volume_extended ocf_composite_volume_extended = {
+	.allocator_type = &ocf_composite_io_allocator_type,
+};
+
+int ocf_composite_volume_type_init(ocf_ctx_t ctx)
+{
+	return ocf_ctx_register_volume_type_internal(ctx,
+			OCF_VOLUME_TYPE_COMPOSITE,
+			&ocf_composite_volume_properties,
+			&ocf_composite_volume_extended);
+}
+
+int ocf_composite_volume_create(ocf_composite_volume_t *volume, ocf_ctx_t ctx)
+{
+	ocf_volume_type_t type;
+
+	type = ocf_ctx_get_volume_type(ctx, OCF_VOLUME_TYPE_COMPOSITE);
+	if (!type)
+		return -OCF_ERR_INVAL;
+
+	return ocf_volume_create(volume, type, NULL);
+}
+
+void ocf_composite_volume_destroy(ocf_composite_volume_t cvolume)
+{
+	ocf_volume_destroy(cvolume);
+}
+
+int ocf_composite_volume_add(ocf_composite_volume_t cvolume,
+		ocf_volume_type_t type, struct ocf_volume_uuid *uuid,
+		void *volume_params)
+{
+	struct ocf_composite_volume *composite = ocf_volume_get_priv(cvolume);
+	ocf_volume_t volume;
+	int result;
+
+	if (composite->members_cnt >= OCF_COMPOSITE_VOLUME_MEMBERS_MAX)
+		return -OCF_ERR_INVAL;
+
+	volume = &composite->member[composite->members_cnt].volume;
+	result = ocf_volume_init(volume, type, uuid, true);
```

Comment:
I can't find a corresponding ocf_volume_destroy() on the subvolumes.

Reply:
Fixed.